### PR TITLE
[WIP] Index File Compression Implementation

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

Partially compress BTree index and Bitmap Index.

## How was this patch tested?

current tests pass

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala [hunk 1]
```diff
@@ -44,6 +42,8 @@ private[index] case class BTreeIndexRecordWriter(
   @transient private lazy val genericProjector = FromUnsafeProjection(keySchema)
   private lazy val nnkw = new NonNullKeyWriter(keySchema)
 
+  private val codecFactory = new CodecFactory(configuration)
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordWriter.scala]: No specific to this change, but maybe we need to add some transient properly in our code

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala [hunk 2]
```diff
@@ -193,4 +194,34 @@ private[oap] object IndexUtils {
       (m, found)
     }
   }
+  def compressIndexData(
+      configuration: Configuration,
+      codecFactory: CodecFactory,
+      bytes: Array[Byte]): Array[Byte] = {
+    val codec = CompressionCodec.valueOf(
+        configuration.get(OapFileFormat.COMPRESSION, OapFileFormat.DEFAULT_COMPRESSION))
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala]: how about just pass the codec or even codec string to this function? That would make this util easier for unit test.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala [hunk 3]
```diff
@@ -193,4 +195,33 @@ private[oap] object IndexUtils {
...
+      val decompressor = codecFactory.getDecompressor(c)
+      val length = Platform.getInt(bytes, Platform.BYTE_ARRAY_OFFSET + 9)
+      val dec = decompressor.decompress(bytes.slice(13, bytes.length), length)
+      dec
+    } else {
+      bytes
+    }
+  }
+
+  def hash(bytes: Array[Byte]): String = {
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala]: Is this used for security check? Which function will call it??

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala (hunk 2 above)
Comment (yueawang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala]: Should we use a OapCodec inherits CompressionCodec? so we can add specific codec on our own.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala (hunk 3 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexUtils.scala]: just for debug if I decompress correctly. I'll remove it.

**Comment by linhong-intel** [comment]
@adrian-wang 

> Not specific to this change, but maybe we need to add some transient properly in our code

Done in this PR. I'll fix other `codecFactory` seperately

> how about just pass the codec or even codec string to this function? That would make this util easier for unit test.

Each open file should have separated `CodecFactory/Compressor/Decompressor`. Since they are not thread-safe

@yueawang 
> Should we use a OapCodec inherits CompressionCodec? so we can add specific codec on our own.

Good idea.  It's also requested by Baidu.

**Comment by linhong-intel** [comment]
Dup to #709 
close this.

**Pull request closed without merging**